### PR TITLE
RavenDB-17888 - Get the order by fields from Lucene

### PR DESCRIPTION
**Pull request by grisha-kotler**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17888

### Additional description

Get the order by fields from Lucene instead of taking them from the stored fields

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs [hunk 1]
```diff
@@ -19,12 +22,26 @@ public partial class IndexReadOperation
 
         foreach (var field in query.Metadata.OrderBy)
         {
-            // https://issues.hibernatingrhinos.com/issue/RavenDB-17888
-            DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Grisha, DevelopmentHelper.Severity.Normal,
-                "make it work without storing the fields in the index");
+            string fieldName;
+            object fieldValue;
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs]: I wonder if we can avoid the boxing here.

File: src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs [hunk 2]
```diff
@@ -2036,7 +2036,18 @@ public static void WriteMetadata(this AbstractBlittableJsonTextWriter writer, Do
...
                     writer.WritePropertyName(orderByField.Field);
-                    writer.WriteString(orderByField.Value);
+                    switch (orderByField.OrderType)
+                    {
+                        case OrderByFieldType.Long:
+                            writer.WriteInteger((long)orderByField.Value);
+                            break;
+                        case OrderByFieldType.Double:
+                            writer.WriteDouble((double)orderByField.Value);
+                            break;
```
Comment (ayende) [inline src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs]: We read it as numeric, then turn that into string here, then read it back again

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs [hunk 3]
```diff
@@ -19,12 +22,26 @@ public partial class IndexReadOperation
...
 
-            var fieldValue = document.GetField(field.Name.Value).StringValue(_state);
-            documentWithOrderByFields.AddOrderByField(field.Name.Value, fieldValue);
+            switch (field.OrderingType)
+            {
+                case OrderByFieldType.Long:
+                    fieldName = $"{field.Name}{Constants.Documents.Indexing.Fields.RangeFieldSuffixLong}";
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs]: This is likely going to create a _lot_ of temporary variables, no? Once for each value in the query

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs [hunk 4]
```diff
@@ -19,12 +22,26 @@ public partial class IndexReadOperation
...
+                    fieldValue = _searcher.IndexReader.GetLongValueFor(fieldName, FieldCache_Fields.NUMERIC_UTILS_LONG_PARSER, doc, _state);
+                    break;
+                case OrderByFieldType.Double:
+                    fieldName = $"{field.Name}{Constants.Documents.Indexing.Fields.RangeFieldSuffixDouble}";
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs]: Same

Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs]: done

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs]: done

File: src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/Persistence/Lucene/IndexReadOperation.Sharding.cs]: done

File: src/Raven.Server/Documents/Replication/Senders/MergedEnumerator.cs [hunk 5]
```diff
@@ -35,7 +35,7 @@ public virtual bool MoveNext()
             var current = _workEnumerators[0];
             for (var index = 1; index < _workEnumerators.Count; index++)
             {
-                if (_comparer.Compare(_workEnumerators[index].Current, current.Current) > 0)
+                if (_comparer.Compare(_workEnumerators[index].Current, current.Current) < 0)
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/Senders/MergedEnumerator.cs:38]: What is this change about? Doesn't seem related to the PR

**Comment by ayende** [comment]
Looks good

File: src/Raven.Server/Documents/Replication/Senders/MergedEnumerator.cs (hunk 5 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/Senders/MergedEnumerator.cs:38]: `MergedEnumerator` is inherited only by the `MergedReplicationBatchEnumerator` which has its own implementation of `MoveNext`. So the `MoveNext` from the `MergedEnumerator` isn't used at all.
I'll also change it in previous versions.

Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/Senders/MergedEnumerator.cs:38]: Actually, it was introduced in 6.0.

File: src/Raven.Server/Documents/Queries/OrderByField.cs [hunk 6]
```diff
@@ -27,6 +34,8 @@ public OrderByField(QueryFieldName name, OrderByFieldType orderingType, bool asc
 
         public readonly MethodType? Method;
 
+        public readonly string LuceneOrderByName;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Queries/OrderByField.cs]: @arekpalinski how this will be in Corax? Can we just call it OrderByName?

Comment (arekpalinski) [reply src/Raven.Server/Documents/Queries/OrderByField.cs]: When using Corax we don't have the notion of `_D`, `_L` fields. For a single numeric field we have string, long and double values. I'm for `OrderByName` or `SearchEngineOrderByName`.

Comment (grisha-kotler) [reply src/Raven.Server/Documents/Queries/OrderByField.cs]: Renamed

**Review by ppekrol — APPROVED** [review]
